analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs

```diff
@@ -50,27 +61,61 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
             if (CheckExpression(equalsExpression.Left, equalsExpression.Right, c.SemanticModel, out var constantValue)
                 || CheckExpression(equalsExpression.Right, equalsExpression.Left, c.SemanticModel, out constantValue))
             {
-                c.ReportIssue(Diagnostic.Create(Rule,
-                                                               equalsExpression.GetLocation(),
-                                                               constantValue < 0 ? "negative" : "positive"));
+                c.ReportIssue(Diagnostic.Create(Rule, equalsExpression.GetLocation(), constantValue < 0 ? "negative" : "positive"));
             }
         }
 
         private static bool CheckExpression(SyntaxNode node, ExpressionSyntax modulus, SemanticModel semanticModel, out int constantValue) =>
             ExpressionNumericConverter.TryGetConstantIntValue(node, out constantValue)
             && constantValue != 0
             && IsModulus(modulus)
-            && !IsUnsigned(modulus, semanticModel);
+            && !ExpressionIsAlwaysPositive(modulus, semanticModel);
```

Comment:
the check `expression.RemoveParentheses() is BinaryExpressionSyntax binary` is done in `IsModulus()`

and  `var leftExpression = ((BinaryExpressionSyntax)expression).Left;` is done inside `ExpressionIsAlwaysPositive`

this makes the code hard to read

you could just inline `IsModulus` and then change the signature of `ExpressionIsAlwaysPositive` to receive `BinaryExpressionSyntax`

like this you would also avoid having the surprise that `expression` is actually a parenthesized expression (you don't have a test but I would expect it would throw when doing the cast in `ExpressionIsAlwaysPositive()`

Reply:
your update matches what I had in mind